pandas/io/stata.py

```diff
@@ -1345,6 +1358,307 @@ def data(self, convert_dates=True, convert_categoricals=True, index=None,
+            cols = np.where(lmap(lambda x: x in _date_formats,
+                                 self.fmtlist))[0]
+            for i in cols:
+                col = data.columns[i]
+                data[col] = _stata_elapsed_date_to_datetime_vec(data[col], self.fmtlist[i])
+
+        if convert_categoricals and self.value_label_dict:
+            data = self.convert_categoricals(data, self.value_label_dict, self.lbllist,
+                                             order_categoricals)
+
+        if not preserve_dtypes:
+            retyped_data = []
+            convert = False
+            for col in data:
+                dtype = data[col].dtype
+                if dtype in (np.float16, np.float32):
+                    dtype = np.float64
+                    convert = True
+                elif dtype in (np.int8, np.int16, np.int32):
+                    dtype = np.int64
+                    convert = True
+                retyped_data.append((col, data[col].astype(dtype)))
+            if convert:
+                data = DataFrame.from_items(retyped_data)
+
+        return data
+
+
+    def convert_categoricals(self, data, value_label_dict, lbllist, order_categoricals):
+        """
+        Converts columns to Categorical type.
+
+        To convert categorical variables in Stata file formats less
+        than 117 when using `get_chunk`, read the file once through to
+        the end to get the categorical label information, then read
+        again and process each chunk with this method.  For format
+        117 or higher the categorical variables can be automatically
+        converted and two passes are not needed.
+
+        Parameters
+        ----------
+        data : DataFrame
+            The data frame to be converted
+        value_label_dict : dict-like
+            Obtained from StataReader object after reading file
+        lbllist : list-like
+            Obtained from StataReader object after reading file
+        order_categoricals: boolean
+            Flag indicating whether converted categorical data are ordered.
+
+        Returns
+        -------
+        DataFrame with columns replaced by categorical variables as specified
+        in the Stata file.
+
+        Notes
+        -----
+        The `value_label_dict` and `lbllist` fields will always be
+        available after the file is completely read, either using the
+        `data` method or repeated calls to the `get_chunk` method.  If
+        the file format is 117 or greater, these attributes will be
+        available immediately upon class initialization.
+
+        Example
+        -------
+        # First pass
+        >> itr = read_stata('file.dta', iterator=True)
+        >> for j in range(500):
+        >>     chunk = itr.get_chunk(10000, convert_categoricals=False)
+        >>     if chunk is None:
+        >>         break
+        >> lbllist = itr.lbllist
+        >> value_label_dict = itr.value_label_dict
+
+        # Second pass
```

Comment:
Would be useful if the second pass could always generate the same `DataFrame` -- dtypes, categoricals, etc, as a 1-pass read.